spring/boot-webflux-autoconfigure/src/main/java/com/linecorp/armeria/spring/web/reactive/ArmeriaHttpClientResponseSubscriber.java

```diff
@@ -74,130 +81,146 @@ public void onSubscribe(Subscription s) {
     }
 
     ResponseBodyPublisher toResponseBodyPublisher() {
-        if (!future.isDone()) {
+        if (!headersFuture.isDone()) {
             throw new IllegalStateException("HTTP headers have not been consumed yet.");
         }
-        final ResponseBodyPublisher publisher = new ResponseBodyPublisher(this);
-        this.publisher = publisher;
-        return publisher;
+        return bodyPublisher;
     }
 
-    final class ResponseBodyPublisher implements Publisher<HttpObject>, Subscription {
+    static final class ResponseBodyPublisher implements Publisher<HttpObject>, Subscription {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<ResponseBodyPublisher, Subscriber> subscriberUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(ResponseBodyPublisher.class, Subscriber.class,
+                                                       "subscriber");
+
+        private final ArmeriaHttpClientResponseSubscriber parent;
 
         @Nullable
-        private Subscriber<? super HttpObject> subscriber;
+        private volatile Subscriber<? super HttpObject> subscriber;
 
-        private final ArmeriaHttpClientResponseSubscriber upstreamSubscriber;
+        @Nullable
+        private Publisher<HttpObject> publisherForLateSubscribers;
 
-        private ResponseBodyPublisher(ArmeriaHttpClientResponseSubscriber upstreamSubscriber) {
-            this.upstreamSubscriber = upstreamSubscriber;
+        ResponseBodyPublisher(ArmeriaHttpClientResponseSubscriber parent) {
+            this.parent = parent;
         }
 
         @Override
         public void subscribe(Subscriber<? super HttpObject> s) {
-            subscriber = s;
-            s.onSubscribe(this);
+            for (;;) {
+                final Subscriber<? super HttpObject> oldSubscriber = subscriber;
+                if (oldSubscriber == null) {
+                    // The first subscriber.
+                    if (subscriberUpdater.compareAndSet(this, null, s)) {
+                        s.onSubscribe(this);
+                        break;
+                    }
+                } else {
+                    // The other subscribers - notify whether completed successfully only.
+                    if (publisherForLateSubscribers == null) {
+                        @SuppressWarnings({ "unchecked", "rawtypes" })
+                        final Publisher<HttpObject> newPublisher =
+                                (Publisher) Mono.fromFuture(parent.completionFuture);
+                        publisherForLateSubscribers = newPublisher;
+                    }
+                    publisherForLateSubscribers.subscribe(s);
+                    break;
+                }
+            }
         }
 
         @Override
         public void request(long n) {
-            if (!isCompleted) {
-                upstreamSubscriber.ensureSubscribed().request(n);
-                return;
+            if (parent.completedCause == null) {
```

Comment:
Is there any chance that this method is called from another thread? If so, I think we can make this volatile.

Reply:
Fixed by making it volatile